streamingvisitors/src/tests/matching_elements_filler/matching_elements_filler_test.cpp

```diff
@@ -0,0 +1,422 @@
+}
+
+void
+MatchingElementsFillerTest::assert_elements(uint32_t doc_lid, const vespalib::string& field, const ElementVector& exp_elements)
+{
+    auto act_elements = _matching_elements->get_matching_elements(doc_lid, field);
+    EXPECT_EQ(exp_elements, act_elements);
+}
+
+void
+MatchingElementsFillerTest::assert_same_element(const vespalib::string& field, const vespalib::string& term1, const vespalib::string& term2, const ElementVector& exp_elements)
+{
+    fill_matching_elements(make_same_element(field, term1, term2));
+    assert_elements(1, field, exp_elements);
+}
+
+void
+MatchingElementsFillerTest::assert_same_element_single(const vespalib::string& field, const vespalib::string& term, const ElementVector& exp_elements)
+{
+    fill_matching_elements(make_same_element_single(field + "." + term));
+    assert_elements(1, field, exp_elements);
+}
+
+TEST_F(MatchingElementsFillerTest, same_element_operator)
+{
+    assert_same_element("elem_array", "name:bar", "weight:20", { 1 });
+    assert_same_element("elem_array", "name:zap", "weight:20", { 4, 5 });
+    assert_same_element("elem_map", "value.name:bar", "value.weight:20", { 0 });
+    assert_same_element("elem_map", "value.name:zap", "value.weight:20", { 4, 5 });
+    assert_same_element("str_int_map", "key:bar", "value:20", { 0 });
+    assert_same_element("str_int_map", "key:zap", "value:20", { 4, 5 });
+}
+
+TEST_F(MatchingElementsFillerTest, nested_field)
+{
+    assert_same_element_single("elem_array", "name:bar", { 1 });
+    assert_same_element_single("elem_array", "name:foo", { 0, 3 });
+    assert_same_element_single("elem_array", "name:zap", { 4, 5 });
+    assert_same_element_single("elem_array", "weight:20", { 1, 4, 5 });
+    assert_same_element_single("elem_map", "key:foo", { 2, 3 });
+    assert_same_element_single("elem_map", "key:zap", { 4, 5 });
+    assert_same_element_single("elem_map", "value.name:bar", { 0 });
+    assert_same_element_single("elem_map", "value.name:foo", { 2, 3 });
+    assert_same_element_single("elem_map", "value.name:zap", { 4, 5 });
+    assert_same_element_single("elem_map", "value.weight:20", { 0, 4, 5 });
+    assert_same_element_single("str_int_map", "key:bar", { 0 });
+    assert_same_element_single("str_int_map", "key:foo", { 2, 3 });
+    assert_same_element_single("str_int_map", "key:zap", { 4, 5 });
+    assert_same_element_single("str_int_map", "value:20", { 0, 4, 5 });
+    assert_same_element_single("str_int_map", "value:10", { 2 });
+}
+
+TEST_F(MatchingElementsFillerTest, traverse_query_tree)
+{
+    MyQueryBuilder builder;
+    builder.addAnd(2);
+    builder.add_term("elem_array.name:bar", 0);
+    builder.make_same_element("elem_map", "value.name:zap", 1, "value.weight:20", 2);
+    fill_matching_elements(make_query(builder.build()));
+    assert_elements(1, "elem_array", { 1 });
+    assert_elements(1, "elem_map", { 4, 5 });
+}
+
+TEST_F(MatchingElementsFillerTest, ignore_same_element_operator)
```

Comment:
Consider being more specific in test name (e.g. and_not_query_node_ignores_all_but_first_child)